src/engine/BMInterfaceTournament.php

```diff
@@ -0,0 +1,1101 @@
+                NULL,
+                0,                   // needs to be non-null, but also a non-player ID
+                TRUE,
+                array(),
+                $tournament->tournamentId,
+                $gameData['roundNumber']
+            );
+
+            // add game number to $this->gameIdArrayArray
+            if (is_array($interfaceResponse) && array_key_exists('gameId', $interfaceResponse)) {
+                $gameIdArrayArray = $tournament->gameIdArrayArray;
+                $gameIdArrayArray[$gameData['roundNumber'] - 1][] = $interfaceResponse['gameId'];
+                $tournament->gameIdArrayArray = $gameIdArrayArray;
+            }
+        }
+    }
+
+    /**
+     * Most of the tournament saving logic
+     *
+     * @param BMTournament $tournament
+     */
+    protected function save_basic_tournament_parameters(BMTournament $tournament) {
+        $query = 'UPDATE tournament '.
+                 'SET status_id = '.
+                 '        (SELECT id FROM tournament_status WHERE name = :status),'.
+                 '    tournament_state = :tournament_state,'.
+                 '    round_number = :round_number '.
+                 'WHERE id = :tournament_id;';
+        $parameters = array(':status' => $this->get_tournament_status($tournament),
+                            ':tournament_state' => $tournament->tournamentState,
+                            ':round_number' => $tournament->roundNumber,
+                            ':tournament_id' => $tournament->tournamentId);
+        self::$db->update($query, $parameters);
+    }
+
+    /**
+     * Logic to save player parameters
+     *
+     * @param BMTournament $tournament
+     */
+    protected function save_player_parameters(BMTournament $tournament) {
+        // james: In adding the ability to save button selection, I've taken the shortcut
+        //        of assuming that only one button is being selected at the moment
+
+        foreach ($tournament->playerIdArray as $playerIdx => $playerId) {
+            $query = 'UPDATE tournament_player_map '.
+                     'SET position = :position,'.
+                     '    remain_count = :remain_count,'.
+                     '    button_id = :button_id '.
+                     'WHERE tournament_id = :tournament_id '.
+                     'AND player_id = :player_id;';
+            $parameters = array(':position' => $playerIdx + 1,
+                                ':remain_count' => $tournament->remainCountArray[$playerIdx],
+                                ':button_id' => $tournament->buttonIdArrayArray[$playerId][0],
+                                ':tournament_id' => $tournament-> tournamentId,
+                                ':player_id' => $playerId);
+            self::$db->update($query, $parameters);
+        }
+    }
+
+    /**
+     * Force a tournament to update
+     *
+     * @param int $tournamentId
+     */
+    protected function update_tournament($tournamentId) {
```

Comment:
At the risk of being a pest: i think this function should be renamed.  `updateFunction` is an API call, so it's pretty confusing for BMInterfaceTournament to contain a function which is named `update_tournament()` and which does not directly implement that API call.  Call it `load_and_save_tournament()` or something.

Reply:
Changed to advance_tournament()